xwayland/xwayland.c

```diff
@@ -169,6 +170,14 @@ static void handle_client_destroy(struct wl_listener *listener, void *data) {
 	struct wlr_xwayland *wlr_xwayland =
 		wl_container_of(listener, wlr_xwayland, client_destroy);
 
+	struct wlr_seat *seat = wlr_xwayland->seat;
+	struct wl_list new_surface_signals;
+	struct wl_listener *new_surface_listener, *next;
+	wl_list_init(&new_surface_signals);
+	wl_list_for_each_safe(new_surface_listener, next, &wlr_xwayland->events.new_surface.listener_list, link) {
+		wl_list_remove(&new_surface_listener->link);
+		wl_list_insert(&new_surface_signals, &new_surface_listener->link);
+	}
```

Comment:
You can do this more simply:

```c
struct wl_list new_surface_signals;
wl_list_init(&new_surface_signals);
wl_list_insert_list(&new_surface_signals, &wlr_xwayland->events.new_surface.listener_list);
wl_list_init(&wlr_xwayland->events.new_surface.listener_list);
```

Same for a few lines below.

I don't really like this, maybe it's possible to add a new function that doesn't reset these fields? (`wlr_xwayland_start`?)

Reply:
I don't like this _at all_ :)
See PR summary comment, I was suggesting something like that yes.

Reply:
Done something better. I think I've addressed all your points with that, feel free to have another look.